### ISSUE TITLE
rewrite: NDI Source 5 2ch+ Distortion

### ISSUE THREAD

**Issue by Riotline** [Palakis/obs-ndi#752]
> @Riotline Can you open a separate issue on what you are seeing?
> _Originally posted by @paulpv in https://github.com/Palakis/obs-ndi/issues/745#issuecomment-1229113058_

Using _NDI™ 5 Source_ source from the rewrite branch in OBS 28.0.0 RC2 appears to work normally with a mono NDI source; however, when using a stereo source, the right channel (ch2) is heavily distorted when pulled in through the _NDI™ 5 Source_ source. Tested through two methods:

- NDI Tools Screen Capture HX to obtain the audio from the PC
- NDI Audio Direct VST3 through Elgato Wave Link Channel. 

Both showed the same distortion on the second channel. Also can confirm that the NDI sources are clean throughout this process. Verified the integrity of the output: 

- NDI™ Tools Studio Monitor to listen to the sources 
- NDI™ Audio Direct via Adobe Audition to ensure multichannel integrity

The _NDI™ Source 5_ source in OBS produced the same multichannel distortion with many permutations of the settings Low Latency, Hardware Acceleration, High Quality, Low Quality, and Audio Only.
Additional experimentation showed that going to higher channel counts (e.g. quad, eight channel) causes distortion on the first channel as well (and potentially the other channels).  Mono operation on any channel except 1 will also produce this distortion.

TLDR: Currently seeing audio distortion of NDI Source audio when the source uses 2ch+. Distortion is not from the actual NDI output but rather the plugin.  

**Specs:**
Windows 11 Build 22000.856
OBS Studio 28.0.0-rc2
NDI v5.5.1

Audio Example of the distortion effect in the stereo configuration: https://github.com/Palakis/obs-ndi/issues/745#issuecomment-1228043019
The example is done through OBS Recording and is consistent with audio produced by audio monitoring of the NDI Source from within OBS.

**Comment by paulpv** [comment]
@Riotline Thanks! I did miss in your [now obvious] comment that this was on the rewrite branch.
I was perusing the previously opened issues and there are several issues about choppiness on Linux or MacOS.
Perhaps there are some clues there.
I did not realize that @tt2468 's rewrite branch was usable/testable yet. I will give it a try myself, hopefully some time this weekend.

**Comment by paulpv** [comment]
@Riotline I built OBS 28 and the obs-ndi rewrite today and tested it on Windows 10 Pro 21H2 19044.1889 in stereo and see/hear no audio distortion:
https://youtu.be/YtD_bmHwvFQ

**Comment by Riotline** [comment]
@paulpv  I wonder if there is something occurring with Windows 11. I'll do tests with my Windows 10 and 11 PCs today and will get back to this.

**Comment by Riotline** [comment]
@paulpv Tested on Windows 10 Pro and Windows 11 Pro and still had the stereo distortion on the second channel.
I'll continue to analyse and experiment with some configurations as well as tinker with the way audio is handled by obs-ndi.

**Comment by tt2468** [comment]
The audio distortion should not be platform related. Just to confirm, is it something easily noticeable, or something that you'd have to pick out? I think perhaps providing audio recordings of current/expected behavior could help a lot.

**Comment by paulpv** [comment]
@Riotline Is this a local loopback from PC1 TXing to PC1 RXing?
I only tested PC1 [OBS28 NDI5] TXing to PC2 [OBS27 NDI4] RXing and then TXing back to PC1.
I will try a more single OBS28 PC loopback.
Can you try OBS28 to OBS27, or OBS27 to OBS28?

**Comment by Riotline** [comment]
> @Riotline Is this a local loopback from PC1 TXing to PC1 RXing? I only tested PC1 [OBS28 NDI5] TXing to PC2 [OBS27 NDI4] RXing and then TXing back to PC1. I will try a more single OBS28 PC loopback. Can you try OBS28 to OBS27, or OBS27 to OBS28?

My first two tests were done solely on one PC running Windows 11 Pro with OBS 28. As mentioned in the original post here, I used two methods of TXing using NDI to ensure integrity.
My more recent test used two PCs running Windows 10 Pro and 11 Pro with OBS 28 on both. Did W11 TXing to W10 RXing. I will try W10 to W11. I'll also try OBS 27 after.

**Comment by paulpv** [comment]
A few ideas I had that woke me up:
1. Can you play a consistent tone sound such as https://www.szynalski.com/tone-generator/ and then record the problem and post the recording here?
Looking at the values recorded when expecting a consistent tone might provide a clue where the problem is.
You could even just generate a 5 second tone in Audacity and Media Source loop it; don't worry about nicely wrapping the end to the beginning, that too is possible to provide a clue.
2. How are you setting up your mono source? Maybe export your active Scene Collection and/or Profile for me to try?
3. Does the problem occur if you only use a stereo source, or does it require you to start with a mono and then toggle to a stereo?
4. What command(s) did you use to build your OBS and obs-ndi? Are you certain you are running a release build?
5. What is your hardware?

I will focus on local only repros and mono to stereo toggling and verifying audio fidelity on all channels.

**Comment by Riotline** [comment]
@paulpv 
1. Attached are two recordings from the Windows 11 Pro PC with tone generation from NDI Tools v5.5.1:
https://user-images.githubusercontent.com/16257629/187070616-79743636-878a-4018-9e09-b698bf6874d7.mp4
https://user-images.githubusercontent.com/16257629/187070614-8427dd29-4d5c-46f8-9943-967ffb2a3c39.mp4

These other two are from the same PC but with music as the NDI Source coming from the Windows 10 PC through the NDI Program Output:
https://user-images.githubusercontent.com/16257629/187070664-02fbb0e4-7141-4a7e-99af-f490976d0ddb.mp4
https://user-images.githubusercontent.com/16257629/187070666-d37e1bcc-eb74-479e-bc4b-b200ee10b179.mp4

2. NDI Tools Audio Direct allows for setting the channels being used on the fly. I was able to set Mono, Stereo, Quad, 8, and 16 on any of the 16 available channels.

3. It has happened even when the source starts as stereo or any higher channel configurations (e.g. eight channel)

4. Used the OBS 28 RC2 release build rather than building in myself
I've used the obs-ndi rewrite artifact from GitHub as well as built it myself on my own branch with various changes in the attempt to fix but same result.

5. 
Windows 11 Pro
Intel i7-11700K
64GB Memory
Nvidia 3060 GPU
ROG Strix Z590-A Mobo
OBS 28.0.0-rc2 x64
OBS 27.2.4 x64

Windows 10 Pro
Intel i9-9900K
32GB Memory
No GPU Right Now
ROG Strix H370-F Mobo
OBS 28.0.0-rc1 (Don't know why i installed rc1) x64

**Comment by paulpv** [comment]
@Riotline You're a rock star for doing all of that.
Interesting that the distortion is much more pronounced in the music than in the tone.
Still looking at it.

**Comment by saschafoerster** [comment]
Sometimes I had problems with audio drivers and different sound cards, when they were using different bitrates. Some soundboards used 44,1khz, others used 48khz. With such setups I also had sometimes distortions (not with NDI) but this is just an idea where to search maybe.

**Comment by paulpv** [comment]
@Riotline I am reproing with using `NDI Screen Capture HX` to output NDI [audio/video] and capture in OBS 28RC2+OBSNDI5rewrite.
My appologies: All of my previous attempts were mostly NDI Outputs only.
I took your original statement:
> Don't know if this would be the right issue thread to add some issues from the rewrite branch but just noticed that NDI output with more than a mono audio channel causes those additional channels to be distorted.
To mean that it is an OBS NDI Output problem.
It does indeed appear to be an OBS NDI Input problem.
I will focus on that.

**Comment by paulpv** [comment]
Repro video https://youtu.be/WNuaYE6M_Z0

In this I am just talking plus stereo music playing.
I do not show mono working fine, nor prove if generated stereo tones work fine or not.

I will look at the NDI Input source.

**Comment by paulpv** [comment]
@tt2468 With the TODO and obviously unfinished code this area is still a WIP, but there are clear differences, mostly in:
1. the hardware acceleration
2. the high performance request
3. dedicated separate audio and video threads

Odd thing is that NDI video seems to be rendering fine even though its code is symmetrical to audio.

`update`
![image](https://user-images.githubusercontent.com/1393897/187090510-1e86f9cb-25f9-4c30-bee7-023838f8d6ae.png)

'audio_thread`
![image](https://user-images.githubusercontent.com/1393897/187090562-7e33280c-4ba0-4eb1-98b6-e7e5eedae71d.png)

**Comment by paulpv** [comment]
@tt2468 Is there a mega benefit to having two threads separately calling `recv_capture_v3`, one for video only, one for audio only, instead of a single call and then processing them both (like the original code)?

**Comment by tt2468** [comment]
@paulpv It is the recommended method by the NDI SDK. Theoretically allows us to enable low latency source mode with low jitter. I don't see a reason why it would affect audio since it's not doing anything with the timestamps where timings could affect it. OBS buffers audio so that it can be continuous. Currently not home to dig deeper myself right now.

**Comment by paulpv** [comment]
@tt2468 I do see this mentioned in the SDK's `Processing.NDI.Recv.h`:
```
// This will allow you to receive video, audio and metadata frames. Any of the buffers can be NULL, in which
// case data of that type will not be captured in this call. This call can be called simultaneously on
// separate threads, so it is entirely possible to receive audio, video, metadata all on separate threads.
// This function will return NDIlib_frame_type_none if no data is received within the specified timeout and
// NDIlib_frame_type_error if the connection is lost. Buffers captured with this must be freed with the
// appropriate free function below.
PROCESSINGNDILIB_API
NDIlib_frame_type_e NDIlib_recv_capture_v3(
	NDIlib_recv_instance_t p_instance,     // The library instance.
	NDIlib_video_frame_v2_t* p_video_data, // The video data received (can be NULL).
	NDIlib_audio_frame_v3_t* p_audio_data, // The audio data received (can be NULL).
	NDIlib_metadata_frame_t* p_metadata,   // The metadata received (can be NULL).
	uint32_t timeout_in_ms                 // The amount of time in milliseconds to wait for data.
);
```
But, no surprise to me, I might have missed it, but do not see any examples of a multi-threaded RX in either the standard or Advanced SDK.

I do see examples of multi-threaded TX in the NDI 5 Advanced SDK's `NDIlib_Send_H264` and `NDIlib_Send_HVEC` codes, where they use a dedicated audio send thread in addition to having the main thread hardware encode and send the video...which may be something obs-ndi may want to try to do.

I am still not clear on the official Advanced SDK Licensing and what is needed to qualify for "NDI|HX3" certification.
I recall seeing an official NDI post/response somewhere briefly explaining the certification process, but I cannot seem to find the mention anywhere right now.

**Comment by paulpv** [comment]
One thing I noticed is:
1. the v4 code
  1. sleeps 100ms
  2. passes 100ms to recv_capture_v3 timeout_in_ms
  3. set `obs_audio_frame.timestamp = (uint64_t)(audio_frame.timestamp * 100);`
2. the v5 code
  1. sleeps 50ms
  2. passes 50ms to recv_capture_v3 timeout_in_ms
  3. set `obs_audio_frame.timestamp = (uint64_t)(audio_frame.timestamp * 100);`

Should the v5 code be setting `timestamp * 50`?

![image](https://user-images.githubusercontent.com/1393897/187093380-3083421c-d08a-40fd-a9c8-ff71d26b1192.png)

**Comment by Riotline** [comment]
@paulpv I happen to have already tested that in my private branch and no difference. It shouldn't matter though since it mainly deals with the timeout period if I'm not wrong. While the timestamp would have to stay the same since that appears to be converting from the ndi time stamp to an OBS time stamp.

**Comment by paulpv** [comment]
@Riotline I tried both the consistent 50ms change and 100ms change with no affect.
I did see that video is using 100ms consistently and requests performance token with a comment that it is process wide implying audio does not need to request a token.

I am going to try a single threaded version similar to the v4 just to see what happens.

**Comment by paulpv** [comment]
Single thread combined A+V @ 100ms also had no effect.
As a data point, it was actually a little worse.


Unrelated, but FYI @tt2468, I have noticed that NDI Output does not reliably start up when OBS starts, and my OBS 28rc2 is crashing when exiting perhaps because it also doesn't stop reliably? (I need to look at the crash log to confirm if the crash is obs-ndi related).  
I tweaked my NightVsKnight fork a bit by calling output_manager->...update right after the output_manager is created, but I have not pushed it yet because I don't know if it is the ideal location to start, and I suspect it is but from looking at how the v4 plugin starts up and stops in a callback event.

**Comment by tt2468** [comment]
The UX for the plugin is not done yet, which includes auto-starting outputs. I'm thinking I'll be focusing on that as #1 since it's required regardless of how the output is implemented.

**Comment by paulpv** [comment]
@tt2468 Sounds good. I will continue to look in to the distortion. I may even reach out to NDI in their forums.

Is it OK to reach out to them saying that I am working on obs-ndi (basically unofficially representing obs-ndi)?

**Comment by paulpv** [comment]
Posted at https://forums.newtek.com/threads/looking-for-help-in-audio-distortion-of-stereo-ndi-receive-in-official-development-of-obs-ndi-upgrade-for-obs-28-ndi-5-5.166639/

Please let me know if I should change the wording any.

**Comment by paulpv** [comment]
Here is an image [from Audition] of the recorded data zoomed in showing recorded gaps:
![image](https://user-images.githubusercontent.com/1393897/187363675-bab14193-ed1f-4d25-a739-44f021d66627.png)

**Comment by paulpv** [comment]
@Riotline Here is a snippet in your 4th .mp4 showing some less pronounced glitches:
![image](https://user-images.githubusercontent.com/1393897/187369152-6dbefe50-704f-456f-9a5c-2f2c7052b9d9.png)

Mine could be worse because of my local compile?
Obviously non-optimized builds would have a harder time keeping up, but I'm not certain why Mono would work and Stereo+ would not.
I still have not taken the time to prove on my setup the Mono works.
I'll do that tomorrow.

**Comment by paulpv** [comment]
FIXED IT! Simple fix. Just doing some last verifications! ...

**Comment by paulpv** [comment]
The below only fixes the consistent crackling/distortion.
I am still seeing some hiccups/silence in the audio a few times every minute or so.
I have confirmed that I am running `Release` build (I was previously running `RelWithDebInfo`.
For the "a few times every minute or so" issue I may play with bumping RX thread priority one level, or a few other tricks.

The fix:
https://github.com/Palakis/obs-ndi/blob/rewrite/src/input.cpp#L249-L250
![image](https://user-images.githubusercontent.com/1393897/187412441-038c4c2f-6a9c-4936-8143-95315c51177d.png)
```C++
for (size_t i = 0; i < channel_count; ++i)
    obs_audio_frame.data[i] = (uint8_t *)(&ndi_audio_frame.p_data[i * ndi_audio_frame.no_samples]);
```
Should be:
```C++
for (size_t i = 0; i < channel_count; ++i)
    obs_audio_frame.data[i] = (uint8_t *)(&ndi_audio_frame.p_data[i * ndi_audio_frame.channel_stride_in_bytes]);
```

If the latter is really what it should be, then it is also incorrect in the `master` code:
https://github.com/Palakis/obs-ndi/blob/rewrite/src/input.cpp#L249-L250
```C++
for (size_t i = 0; i < channel_count; ++i)
    obs_audio_frame.data[i] = (uint8_t *)(&ndi_audio_frame.p_data[i * ndi_audio_frame.no_samples]);
```
Should be:
```C++
for (size_t i = 0; i < channel_count; ++i)
    obs_audio_frame.data[i] = (uint8_t *)(&ndi_audio_frame.p_data[i * ndi_audio_frame.channel_stride_in_bytes]);
```

I am not 100% sure if `obs_audio_frame.frames = ndi_audio_frame.no_samples;` is OK as is or should be `obs_audio_frame.frames = ndi_audio_frame.channel_stride_in_bytes;`.
I suspect that it is fine as is and supposed to be `no_samples`.

**Comment by tt2468** [comment]
Looks like a simple typo that has managed to exist since day one. Presumably was never bad enough to raise enough eyebrows. Good catch! Using `channel_stride_in_bytes` should make for correct audio interleaving. Go ahead and open a PR to push to `rewrite` and I'll do some testing.

**Comment by Riotline** [comment]
@tt2468 @paulpv 
I added this change to my private testing clone of the rewrite and can confirm that the distortion is not present anymore with this change. Will continue testing

**Comment by paulpv** [comment]
@Spo-oky OBS 27 or 28? You would have to build your own private branch until @tt2468 approves my PR.

@Riotline ... is this you? https://www.youtube.com/c/Riotline/videos :)

**Comment by Riotline** [comment]
> @Riotline ... is this you? https://www.youtube.com/c/Riotline/videos :)
@paulpv indeed me 💀. Stop snooping haha 🤣

After I finish more rounds of testing with this change in the audio thread input and it gets added to the rewrite branch, I shall close it. But for the meantime, will continue testing!

**Comment by Spo-oky** [comment]
Hi,

I'm on OBS 27, I was receiving errors starting OBS 28 rc2 with NDI plugin, so I was waiting for a new plugin. I resolved to update NDI to the last version and all happen.

I've started to create a build directory but I was stuck, I was stuck on QT. lol
I'm not a c++ developer unfortunately, because there are a lot of good projects I will love to help but mainly are on c++ like OBS, CasparCG and others.

Thanks guys.

**Comment by paulpv** [comment]
@Spo-oky check out my build instructions at my fork https://github.com/NightVsKnight/obs-ndi/blob/rewrite_readme/README.md

**Comment by Spo-oky** [comment]
@paulpv  first of all thanks for the help.

I know that this is stupid and it is a simple thing, but it give me an error of command **Build** not found.
I have cmake and VS 2022 installed, what is missing?

Tks

**Comment by paulpv** [comment]
@Spo-oky Building OBS or obs-ndi?
You do need to make sure you have the pre-reqs mentioned at:
https://github.com/obsproject/obs-studio/wiki/build-instructions-for-windows
(I will update my obs-ndi readme)

Can you copy and paste the error output here?

**Comment by Spo-oky** [comment]
Hi,

Sorry about the delay, to many work today. I've deleted all directory and started over.

I managed to pass the first error, that was on OBS build.
I added the dependencies to your commands:

before `PS …\obs-studio> .\CI\build-windows.ps1`
I've added `CI/windows/01_install_depen
[CMakeOutput.log](https://github.com/Palakis/obs-ndi/files/9464912/CMakeOutput.log)
dencies.ps1`

Now on the NDI I've copied the `.github\scripts` directory from the original and buildspec.json.

When I execute `.\.github\scripts\Build-Windows.ps1` I received the error 


```
-- Could NOT find Libobs (missing: LIBOBS_LIB LIBOBS_INCLUDE_DIR)
CMake Error at external/FindLibObs.cmake:106 (message):
  Could not find the libobs library
Call Stack (most recent call first):
  CMakeLists.txt:17 (include)

-- Configuring incomplete, errors occurred!
See also "D:/OBSNDIFix/obs-ndi/build_x64/CMakeFiles/CMakeOutput.log".
Exception: D:\OBSNDIFix\obs-ndi\.github\scripts\Build-Windows.ps1:101
Line |
 101 |  Build
     |  ~~~~~
     | cmake -S . -B build_x64 -G Visual Studio 17 2022 -DCMAKE_SYSTEM_VERSION=10.0.20348.0
     | -DCMAKE_GENERATOR_PLATFORM=x64 -DCMAKE_BUILD_TYPE=RelWithDebInfo
     | -DCMAKE_PREFIX_PATH:PATH=D:\OBSNDIFix\obs-build-dependencies\plugin-deps-2022-08-02-qt6-x64
     | -DQT_VERSION=6 exited with non-zero code 1.
```

**Comment by Spo-oky** [comment]
Hi,

I've already figured out, yesterday night after several tries, every time I've made a clone of obs-ndi the version was the master. What have some things missing. After switching to rewrite every thing went ok.

Just doing a live transmission at this time and after I will test it with the new version and give a feedback.

Tks

**Comment by Spo-oky** [comment]
I'm using now the newest version o OBS v28

After this obs-ndi changes I can confirm that the sound are perfect! I'm going to continue the tests because we use it 24x7.

**Comment by Spo-oky** [comment]
Hi, I can confirm that the audio is perfect! The only thing that I found is the it's missing the sync tab on this new plugin.
We had to change the sync to source timing, if not the video was not fluid. It appears that is using a title more computer work and the image is not fluid. Going to make more tests.
